compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/hotspot/aarch64/shenandoah/AArch64HotSpotShenandoahBarrierSetLIRGenerator.java

```diff
@@ -0,0 +1,133 @@
+        Value result = tool.newVariable(tool.toRegisterKind(kind));
+        ForeignCallLinkage callTarget = getReadBarrierStub(tool, strength, narrow);
+        AllocatableValue object = tool.asAllocatable(obj);
+        AArch64AddressValue loadAddress = ((AArch64LIRGenerator) tool).asAddressValue(address, AArch64Address.ANY_SIZE);
+        tool.getResult().getFrameMapBuilder().callsMethod(callTarget.getOutgoingCallingConvention());
+        tool.append(new AArch64HotSpotShenandoahLoadRefBarrierOp(config, providers, tool.asAllocatable(result), object, loadAddress, callTarget, strength, notNull));
+        return result;
+    }
+
+    @Override
+    public void emitPreWriteBarrier(LIRGeneratorTool lirTool, Value address, AllocatableValue expectedObject, boolean nonNull) {
+        AllocatableValue temp = lirTool.newVariable(LIRKind.value(AArch64Kind.QWORD));
+        // If the assembly must load the value then it needs a temporary to store it.
+        AllocatableValue temp2 = expectedObject.equals(Value.ILLEGAL) ? lirTool.newVariable(LIRKind.value(AArch64Kind.QWORD)) : Value.ILLEGAL;
+
+        // Load the address into a register
+        AllocatableValue addressValue = lirTool.newVariable(address.getValueKind());
+        lirTool.emitMove(addressValue, address);
+
+        ForeignCallLinkage callTarget = lirTool.getForeignCalls().lookupForeignCall(HotSpotHostForeignCallsProvider.SHENANDOAH_WRITE_BARRIER_PRE);
+        lirTool.getResult().getFrameMapBuilder().callsMethod(callTarget.getOutgoingCallingConvention());
+        lirTool.append(new AArch64HotSpotShenandoahSATBBarrierOp(config, providers, addressValue, expectedObject, temp, temp2, callTarget, nonNull));
+    }
+
+    @Override
+    public void emitCardBarrier(LIRGeneratorTool lirTool, Value address) {
+        AArch64AddressValue addr = ((AArch64LIRGenerator) lirTool).asAddressValue(address, AArch64Address.ANY_SIZE);
+        AllocatableValue tmp = lirTool.newVariable(LIRKind.value(AArch64Kind.QWORD));
+        lirTool.append(new AArch64HotSpotShenandoahCardBarrierOp(config, providers, addr, tmp));
+    }
+
+    @Override
+    public void emitCompareAndSwapOp(LIRGeneratorTool tool, boolean isLogic, Value address, MemoryOrderMode memoryOrder, AArch64Kind memKind, Variable result,
+                    AllocatableValue allocatableExpectedValue, AllocatableValue allocatableNewValue, BarrierType barrierType) {
+        AllocatableValue tmp1 = tool.newVariable(LIRKind.value(AArch64Kind.QWORD));
+        tool.append(new AArch64HotSpotShenandoahCompareAndSwapOp(config, providers, memKind, memoryOrder, isLogic, result, allocatableExpectedValue, allocatableNewValue, tool.asAllocatable(address),
+                        tmp1));
+    }
+
+    @Override
+    public Value emitAtomicReadAndWrite(LIRGeneratorTool tool, LIRKind readKind, Value address, Value newValue, BarrierType barrierType) {
+        // We insert the necessary barriers in the node graph, at that level it
```

Comment:
minor: this comment is a bit mislead IMHO - I guess it means the necessary barriers have been added to GraalIR in the frontend of the compiler already?